clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
The change I'm suggesting is fully compatible with KafkaConsumer's poll() interface. It returns data when data is available. If the current behavior blocks with data available, then it is not following the API. Fixing this issue here is purely for convenience since it also resolves the same issue you have with unnecessary blocking in the current patch.

If we want to ensure that NetworkClient.poll is always invoked, you could change pollOnce to include this snippet instead:
```java
Map records = fetcher.fetchedRecords();
if (!records.isEmpty()) {
  client.poll(0);
  return records;
} else {
  fetcher.initFetches(cluster);
  client.poll(timeout);
  return fetcher.fetchedRecords();
}
```
(By the way, I'm not a committer, but I did write pollOnce if that's worth anything.)

I don't have a strong preference for this solution, so if you can think of another, I encourage you to implement it. Probably the initial approach would get -1 from me since it makes the fetch response handling more convoluted. Have you talked with @onurkaraman and @becketqin to get more opinions?